File: src/mirv.py
from .entity import Entity
from .trail import Trail
from .shell import Shell
from .inifile import ReadIniFile

class Mirv(Entity):
    OPTION_Fragments = 5
    OPTION_Spread = 0.2

    def __init__(self, game, player, x_launch, y_launch, x_launch_vel, y_launch_vel, launch_time, size, damage):
        super().__init__(game)
        self._player = player
        self._x_launch = x_launch
        self._y_launch = y_launch
        self._x_launch_vel = x_launch_vel
        self._y_launch_vel = y_launch_vel
        self._launch_time = launch_time
        self._size = size
        self._damage = damage
        
        self._x = x_launch
        self._y = y_launch
        
        self._trail = Trail(game, self._x, self._y)
        self._game.add_entity(self._trail)
        
        # Apex calculation
        # y = y0 + vy*t - 5t^2. Apex when vy - 10t = 0 => t = vy/10.
        # But wait, original code says: _apexTime = _launchTime + _yLaunchVel / 10.0f;
        # Valid assumption for gravity=10.
        self._apex_time = launch_time + y_launch_vel / 10.0

    @staticmethod
    def read_settings(settings: ReadIniFile):
        Mirv.OPTION_Fragments = settings.get_int("Mirv", "Fragments", 5)
        Mirv.OPTION_Spread = settings.get_float("Mirv", "Spread", 0.2)

    def draw(self):
        # Draw as tiny triangle (same as Shell)
        if not self._game.get_interface(): return
        
        p1 = self._game.get_interface().game_to_screen(self._x + 0.00, self._y + 0.018)
        p2 = self._game.get_interface().game_to_screen(self._x + 0.03, self._y - 0.018)
        p3 = self._game.get_interface().game_to_screen(self._x - 0.03, self._y - 0.018)
        
        import pygame
        pygame.draw.polygon(self._game.get_interface()._window, (255, 255, 255), [p1, p2, p3])

    def update(self, time):
        old_x = self._x
        old_y = self._y
        
        current_time = self._game.get_time()
        
        if current_time > self._apex_time:
            # Split
            split_time = self._apex_time - self._launch_time
            
            # Position at split
            self._x = split_time * self._x_launch_vel + self._x_launch
            self._y = split_time * (self._y_launch_vel - 5.0 * split_time) + self._y_launch
            
            for i in range(Mirv.OPTION_Fragments):
                # Spread logic
                # fragXvel = _xLaunchVel + (_xLaunchVel * OPTION_Spread * (i - ((Frag-1)/2.0)))
                offset = float(i) - ((Mirv.OPTION_Fragments - 1) / 2.0)
                frag_x_vel = self._x_launch_vel + (self._x_launch_vel * Mirv.OPTION_Spread * offset)
                
                fragment = Shell(self._game, self._player, self._x, self._y,
                                 frag_x_vel, 0.0, self._apex_time,
                                 self._size, self._damage, False)
                self._game.add_entity(fragment)
                
            self._trail.lay_trail(self._x, self._y)
            self._trail.set_inactive()
            return False # Consume MIRV
            
        # Normal update
        time_since_launch = current_time - self._launch_time
        self._x = time_since_launch * self._x_launch_vel + self._x_launch
        self._y = time_since_launch * (self._y_launch_vel - 5.0 * time_since_launch) + self._y_launch
        
        landscape = self._game.get_landscape()
        w = landscape.get_landscape_width() if landscape else 10.0
        
        if self._x > w or self._x < -w:
            self._trail.lay_trail(self._x, self._y)
            self._trail.set_inactive()
            return False
            
        if landscape:
            coll = landscape.ground_collision(old_x, old_y, self._x, self._y)
            if coll[0]:
                self._trail.lay_trail(coll[1], coll[2])
                self.explode(coll[1], coll[2], -1)
                return False
                
        self._trail.lay_trail(self._x, self._y)
        
        players = self._game.get_players()
        for i, p in enumerate(players):
            if p and p.get_tank():
                if p.get_tank().intersect_tank(old_x, old_y, self._x, self._y):
                    self.explode(self._x, self._y, i)
                    return False
        
        return True

    def explode(self, x, y, hit_tank):
        self._trail.set_inactive()
        self._game.explosion(x, y, self._size, self._damage, hit_tank, 1, False, self._player)
        return False

File: groundfire-0.25/src/mirv.cc
////////////////////////////////////////////////////////////////////////////////
//
//               Groundfire
//
////////////////////////////////////////////////////////////////////////////////
//
// Copyright (c) 2004, Tom Russell (tom@groundfire.net)
//
// This file is part of the Groundfire project, distributed under the MIT 
// license. See the file 'COPYING', included with this distribution, for a copy
// of the full MIT licence.
//
////////////////////////////////////////////////////////////////////////////////
//
//   File name : mirv.cc
//
//          By : Tom Russell
//
//        Date : 03-Apr-04
//
// Description : Handles the mirv shells (the ones that split up)
//
//
//
//
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
// Includes
////////////////////////////////////////////////////////////////////////////////
#include "mirv.hh"
#include "trail.hh"
#include "landscape.hh"
#include "shell.hh"
#include "player.hh"

////////////////////////////////////////////////////////////////////////////////
// Define Static Member Variables
////////////////////////////////////////////////////////////////////////////////

int   cMirv::OPTION_Fragments;
float cMirv::OPTION_Spread;

////////////////////////////////////////////////////////////////////////////////
// Public Member Functions
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
//
// Function    : cMirv
//
// Description : Constructor
//
////////////////////////////////////////////////////////////////////////////////
cMirv::cMirv
(
    cGame   * game,
    cPlayer * player,
    float xLaunch,
    float yLaunch,
    float xLaunchVel,
    float yLaunchVel,
    float launchTime,
    float size,    // The size of the blast created when the shell explodes
    float damage   // Amount of damage a direct hit from the shell will do 
)
: cEntity (game),
  _player (player),
  _xLaunchVel (xLaunchVel),
  _yLaunchVel (yLaunchVel),
  _launchTime (launchTime),
  _size (size),
  _damage (damage)
{
    _x = _xLaunch = xLaunch;
    _y = _yLaunch = yLaunch;

    // Create a new trail object to leave behind the shell
    _trail = new cTrail (game, _x, _y);
    _game->addEntity (_trail);

    // Calculate the exact time we'll reach the apex. This is when the mirv
    // will split.
    _apexTime = _launchTime + _yLaunchVel / 10.0f;
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : ~cMirv
//
// Description : Destructor
//
////////////////////////////////////////////////////////////////////////////////
cMirv::~cMirv
(
)
{
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : readSettings
//
// Description : Read the settings for the mirv entities
//
////////////////////////////////////////////////////////////////////////////////
void 
cMirv::readSettings
(
    cReadIniFile const & settings
)
{
    // The number of fragments the mirv will split into.
    OPTION_Fragments
        = settings.getInt ("Mirv", "Fragments", 5);

    // How widely the fragments spread from their parent's trajectory
    OPTION_Spread
        = settings.getFloat ("Mirv", "Spread", 0.2f);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : draw
//
// Description : Draw the shell.
//
////////////////////////////////////////////////////////////////////////////////
void
cMirv::draw
(
    void
)
{
    glLoadIdentity ();

    glTranslatef (_x, _y, -6.0f);

    glColor3f (1.0f, 1.0f, 1.0f);

    // Shells are currently drawn as tiny triangles (they look like dots)
    glBegin (GL_TRIANGLES);
    glVertex3f ( 0.00f,  0.018f, 0.0f);
    glVertex3f ( 0.03f, -0.018f, 0.0f);
    glVertex3f (-0.03f, -0.018f, 0.0f);
    glEnd ();
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : update
//
// Description : Update the mirv
//
////////////////////////////////////////////////////////////////////////////////
bool
cMirv::update
(
    float time
)
{
    float oldX = _x;
    float oldY = _y;

    float collisionX;
    float collisionY;

    float currentTime = _game->getTime ();

    if (currentTime > _apexTime) 
    {
        // It's time to split the mirv up.

        float splitTime = _apexTime - _launchTime;

        // Work out the x,y coordinates of the apex (i.e. where the split 
        // happens.)
        _x = splitTime * _xLaunchVel + _xLaunch;
        _y = splitTime * (_yLaunchVel - 5.0f * splitTime) + _yLaunch;

        for (int i = 0; i < OPTION_Fragments; i++)
        {
            // Calculate the new xVelocity of this fragment.
            float fragXvel = _xLaunchVel + (_xLaunchVel * OPTION_Spread * 
                                            ((float)i - 
                                             ((OPTION_Fragments - 1) / 2.0f)));

            // Create a standard shell for each fragment.
            cShell * fragment = new cShell 
            (
                _game, 
                _player,
                _x,
                _y,
                fragXvel,
                0.0f, 
                _apexTime,
                _size, 
                _damage,
                false
            );
            _game->addEntity (fragment); 
        }

        // Now that the mirv has split, we don't need it anymore so destroy it.
        _trail->layTrail (_x, _y);
        _trail->setInactive ();

        delete this;
        return (false);
    }

    // Calculate the new position of the mirv using a formula based on the 
    // elapsed time since the mirv was fired. The advantage of this is that it
    // is completely independent of the game's frame rate.

    // Get the number of seconds passed since the shell was created.
    float timeSinceLaunch = _game->getTime () - _launchTime;

    _x = timeSinceLaunch * _xLaunchVel + _xLaunch;
    _y = timeSinceLaunch * (_yLaunchVel - 5.0f * timeSinceLaunch) + _yLaunch;

    // If the mirv has left the side of the screen.
    if (_x >  (_game->getLandscape ())->getLandscapeWidth () ||
        _x < -(_game->getLandscape ())->getLandscapeWidth ()) 
    {

        // Lay a final piece of trail and stop laying trail.
        _trail->layTrail (_x, _y);
        _trail->setInactive ();

        // mirv has left the side of the screen so destroy it
        delete this;
        return (false);
    }

    // Have we hit the ground?
    if ((_game->getLandscape ())->groundCollision (oldX, oldY, _x, _y, 
                                                   &collisionX, 
                                                   &collisionY))
    {       
        _trail->layTrail (collisionX, collisionY);

        explode (collisionX, collisionY, -1);

        return (false);
    }

    _trail->layTrail (_x, _y);

    // Check for collision with a tank
    cPlayer ** players = _game->getPlayers ();

    for (int i = 0; i < 8 && players[i] != NULL; i++)
    {
        if (players[i]->getTank ()->intersectTank (oldX, oldY, _x, _y))
        {
            explode (_x, _y, i);

            return (false);
        }
    }
    
    return (true);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : explode
//
// Description : Explode the mirv at the coordinates provided
//
////////////////////////////////////////////////////////////////////////////////
void
cMirv::explode
(
    float x, 
    float y,
    int   hitTank // The tank to deal full damage to (i.e. the tank we hit)
)
{ 
    _trail->setInactive ();
    
    _game->explosion (x, y,
                      _size,
                      _damage,
                      hitTank, 
                      1,
                      false,
                      _player);

    delete this;
}
